data/depreciation_rates/tax_depreciation_rules.json

```diff
@@ -0,0 +1,1181 @@
+                            "RD92", "AE10", "AE20", "AE30", "AE40", "AE50",
+                            "RES", "LAND", "INV"
+                            ]
+                      }
+                  }
+            },
+            "minor_asset_group": {
+              "type": "str",
+              "validators": {
+                      "choice": {
+                          "choices": [
+                              "Computers and Software",
+                              "Industrial Machinery",
+                              "Instruments and Communications Equipment",
+                              "Intellectual Property",
+                              "Inventories",
+                              "Land",
+                              "Mining and Drilling Structures",
+                              "Nonresidential Buildings",
+                              "Office and Residential Equipment",
+                              "Other Industrial Equipment",
+                              "Other Structures",
+                              "Residential Buildings",
+                              "Service industry machinery",
+                              "Transportation Equipment"
+                            ]
+                      }
+                  }
+                },
+            "major_asset_group": {
+              "type": "str",
+              "validators": {
+                      "choice": {
+                          "choices": [
+                              "Equipment",
+                              "Structures",
+                              "Intellectual Property",
+                              "Inventories",
+                              "Land"
+                            ]
+                      }
+                  }
+                },
+            "ADS_life": {
+              "type": "float",
+              "validators": {
+                      "choice": {
+                          "choices": [3.0, 4.0, 5.0, 6.0, 7.0, 9.0, 9.5, 10.0,
+                                      12.0, 14.0, 15.0, 18.0, 19.0, 20.0, 25.0,
+                                      28.0, 30.0, 40.0, 50.0, 100.0]
+                      }
+                  }
+                },
+            "GDS_life": {
+              "type": "float",
+              "validators": {
+                      "choice": {
+                          "choices": [3.0, 5.0, 7.0, 10.0, 15.0, 20.0, 25.0,
+                                      27.5, 39.0, 100.0]
+                      }
+                  }
+                },
+            "system": {
+              "type": "str",
+              "validators": {
+                      "choice": {
+                          "choices": ["ADS", "GDS", "Economic"]
+                      }
+                  }
+                },
+            "year": {
+                "type": "int",
+                "validators": {"range": {"min": 2013, "max": 2030}}
+            },
+            "life": {
```

Comment:
@hdoupe I'm having some trouble setting up validators for `life` and `method`.  I'd like to avoid specifying them for each asset, but setting the validation here don't seem to work (though there appears to be some validation going on for the `method` attribute - it just doesn't seem to be coming from the list in this JSON).  See the last two tests in `tests/test_parameters.py` - first one of those passes, second fails.  

Any advice on how to do this?

Reply:
@jdebacker since this is a custom field, you have to set them in the Python code:

```python
class DepreciationRules(ma.Schema):
    # currently no max is set
    life = ma.fields.Integer(validate=ma.validate.Range(min=0, max=100))
    method = ma.fields.String(
        validate=ma.validate.OneOf(choices=[
            "SL", "Expensing", "DB 150%", "DB 200%", "Economic"])
    )
```

You may want to change life to the `Float` field instead of being an `Integer` field to match your JSON. https://marshmallow.readthedocs.io/en/latest/marshmallow.fields.html#marshmallow.fields.Float

Reply:
@hdoupe Got it - thanks!  Guess I hadn't noticed that one validator in the code already, ah!